File: common/src/main/java/net/impleri/playerskills/network/NetHandler.java
package net.impleri.playerskills.network;

import dev.architectury.networking.simple.MessageType;
import dev.architectury.networking.simple.SimpleNetworkManager;
import net.impleri.playerskills.PlayerSkills;
import net.impleri.playerskills.server.ServerApi;
import net.impleri.playerskills.server.events.SkillChangedEvent;
import net.minecraft.client.player.LocalPlayer;
import net.minecraft.server.level.ServerPlayer;

public abstract class NetHandler {
    public static SimpleNetworkManager NET = SimpleNetworkManager.create(PlayerSkills.MOD_ID);

    public static MessageType SYNC_SKILLS = NET.registerS2C("sync_skills", SyncSkillsMessage::new);

    public static MessageType RESYNC_SKILLS = NET.registerC2S("resync_skills", ResyncSkillsMessage::new);

    public static void syncPlayer(SkillChangedEvent<?> event) {
        var player = event.getPlayer();

        if (player instanceof ServerPlayer serverPlayer) {
            syncPlayer(serverPlayer);
        } else {
            PlayerSkills.LOGGER.warn("Attempted to sync skill changes from clientside");
        }
    }

    public static void syncPlayer(ServerPlayer player) {
        if (player == null) {
            PlayerSkills.LOGGER.warn("Attempted to sync skill changes for nobody");
            return;
        }

        var skills = ServerApi.getAllSkills(player);
        PlayerSkills.LOGGER.debug("Syncing {} player skills to {}", skills.size(), player.getName().getString());

        new SyncSkillsMessage(player, skills).sendTo(player);
    }

    public static void resyncPlayer(LocalPlayer player) {
        PlayerSkills.LOGGER.debug("Requesting skills resync for {}", player.getName().getString());

        new ResyncSkillsMessage(player).sendToServer();
    }
}

File: common/src/main/scala/net/impleri/playerskills/server/NetHandler.scala
package net.impleri.playerskills.server

import net.impleri.playerskills.events.SkillChangedEvent
import net.impleri.playerskills.network.SyncSkillsMessageFactory
import net.impleri.playerskills.server.api.Player
import net.impleri.playerskills.utils.PlayerSkillsLogger
import net.impleri.slab.entity.{Player => MinecraftPlayer}
import net.impleri.slab.logging.Logger

import scala.util.chaining.scalaUtilChainingOps

class NetHandler(
  private val playerOps: Player,
  private val messageFactory: SyncSkillsMessageFactory,
  private val logger: Logger,
) {
  def syncPlayer(player: MinecraftPlayer[_], force: Boolean = true): Unit = {
    playerOps.get(player)
      .tap(logger.debugP(s => s"Syncing ${s.size} player skills to ${player.name}"))
      .pipe(messageFactory.send(player, _, force))
      .foreach(player.sendMessage)
  }

  def syncPlayer(event: SkillChangedEvent[_]): Unit = {
    if (!event.player.isServer) {
      logger.warn("Attempted to sync skill changes from clientside")
    } else {
      syncPlayer(event.player, force = false)
    }
  }
}

object NetHandler {
  def apply(
    playerOps: Player = Player(),
    messageFactory: SyncSkillsMessageFactory = SyncSkillsMessageFactory(),
    logger: Logger = PlayerSkillsLogger.SKILLS,
  ): NetHandler = {
    new NetHandler(
      playerOps,
      messageFactory,
      logger,
    )
  }
}
